File: python-basic/04_loop.py
# 04: 반복문
# 학습 목표: 반복문 종류와 기본 문법 이해, 반복 제어 연습, 컬렉션 순회 연습
# 개념:
# 1. 반복문 종류:
#    - for문: 지정한 범위 내에서 반복, 시퀀스(리스트, 문자열 등) 순회 가능
#    - while문: 조건이 참인 동안 반복 실행
# 2. break/continue: 반복 제어용 키워드로 반복문 탈출 또는 다음 반복으로 건너뜀
# 3. 반복문과 함수 결합: 반복 과정 출력용 함수 호출

# 반복 변수 값 출력 함수 정의
def log_var(var_name, value):
    print(f"변수 {var_name} 값 === {value}")

# for문 예시: 1부터 5까지 출력
for i in range(1, 6):
    log_var("i", i)

print()

# while문 예시: 1부터 5까지 출력
j = 1
while j <= 5:
    log_var("j", j)
    j += 1

print()

# break와 continue 예시
for x in range(1, 11):
    if x % 2 == 0:
        continue  # 짝수일 때는 출력하지 않고 다음 반복으로
    if x > 7:
        break  # 7보다 크면 반복 종료
    log_var("x(홀수 && <=7)", x)

print()

# 리스트 순회를 위한 for문
fruits = ['사과', '바나나', '체리']
for index, fruit in enumerate(fruits):
    log_var(f"fruits[{index}]", fruit)

print()

# 문자열 순회
word = "Python"
for char in word:
    log_var("char", char)

print()

# while문과 조건 결합 예시
count = 0
while True:
    count += 1
    if count > 3:
        break
    log_var("count", count)

File: js-basic/04-loop.js
// 04: 반복문
// 학습 목표: 반복문 종류/문법 익히기, 반복 제어, 배열/객체 순회
// 개념:
// 1. 반복문 종류
//    - for: 반복 횟수가 정해진 경우 (초기값, 조건식, 증감식)
//    - while: 조건이 true인 동안 반복
//    - do-while: 최소 한 번 실행 후 조건에 따라 반복
//    - forEach: 배열의 각 요소별 콜백 실행
//    - for...of: 배열, 문자열 등 iterable 자료 순회 (값 접근)
//    - for...in: 객체의 key 순회
//
// 2. 반복문 내 변수 값 변경, break/continue, 배열/객체 순회 연습

// log 함수: 변수명과 값 출력
const logVar = (varName, value) => {
  console.log(`${varName} 값 === ${value}`);
};

// for문: 1~5 출력
for (let i = 1; i <= 5; i++) {
  logVar("[for문] i", i);
}

// while문: 1~5 출력
let j = 1;
while (j <= 5) {
  logVar("[while문] j", j);
  j++;
}

// do-while문: 최소 1회 실행 후 조건 검사
let k = 1;
do {
  logVar("[do while문] k", k);
  k++;
} while (k <= 5);

// 배열 순회: forEach
const fruits = ["Apple", "Banana", "Cherry"];
fruits.forEach((fruit, idx) => {
  logVar(`fruits[${idx}]`, fruit);
});

// 배열 순회: for...of
for (const fruit of fruits) {
  logVar("fruit(for...of)", fruit);
}

// 객체 순회: for...in
const person = { name: "Alice", age: 30 };
for (const key in person) {
  logVar(`person.${key}`, person[key]);
}

// 반복문 값 변경 예시
let sum = 0;
for (let n = 1; n <= 5; n++) {
  sum += n;
  logVar("sum(누적)", sum);
}

// 반복 제어: break/continue
for (let x = 1; x <= 10; x++) {
  if (x % 2 === 0) continue; // 짝수 건너뜀
  if (x > 7) break; // 8 이상은 반복 종료
  logVar("x(홀수/7이하)", x);
}
